File: rag/advanced_rag/harness/structure_qa.py
"""Ask the chat model to answer a question from a compiled-structure outline.

Shared by the two navigation paths that read compiled rows:

* document structure navigation (``navigation._navigate_within_doc``) — the
  document's catalog / mindmap outline;
* knowledge-graph exploration (``navigation.graph_explore``) — the compiled
  knowledge graph.

Both render entities + relations into the same compact outline, ask the model
whether that outline alone answers the question, and use the returned
``relevant_entities`` to pull the underlying source chunks even when the outline
is NOT sufficient (so evidence still flows back to the caller).

Lives apart from ``navigation`` because it is pure prompt/render work with no
store access, and because both callers would otherwise import it from a module
that in turn imports them.
"""

import logging
import re

import json_repair

_LOG = logging.getLogger(__name__)

# Cap how much compiled structure we render into the prompt.
_MAX_ENTITIES = 300
_MAX_RELATIONS = 300

_NAV_SYSTEM = """You are given {noun} of one or more documents — an outline of entities and their relations — and a question.

Decide whether that outline alone already answers the question.

Rules:
1. Answer ONLY from the outline below. Do not invent facts.
2. Set "is_sufficient" to true only when the outline genuinely answers the question; otherwise false with an empty answer.
3. Always fill "relevant_entities" with the exact `name` values of the entities most related to the question (up to 10), even when the outline is not sufficient — they are used to pull the underlying source text.

Output ONLY JSON, no prose, no code fences:
{{"is_sufficient": true/false, "answer": "<answer, or empty>", "relevant_entities": ["<entity name>", ...]}}"""


def _render_structure(entities: list[dict], relations: list[dict]) -> str:
    """Render the compiled structure as a compact outline for the prompt."""
    lines: list[str] = []
    if entities:
        lines.append("Entities:")
        for e in entities[:_MAX_ENTITIES]:
            name = (e.get("name") or "").strip()
            if not name:
                continue
            typ = (e.get("type") or "other").strip()
            desc = " ".join((e.get("description") or "").split())
            lines.append(f"- {name} ({typ})" + (f": {desc}" if desc else ""))
    if relations:
        lines.append("\nRelations:")
        for r in relations[:_MAX_RELATIONS]:
            src, tgt = (r.get("from") or "").strip(), (r.get("to") or "").strip()
            if not src or not tgt:
                continue
            lines.append(f"- {src} -[{(r.get('type') or 'related').strip()}]-> {tgt}")
    return "\n".join(lines)


async def _ask_structure(tools, topic: str, entities: list[dict], relations: list[dict], noun: str, label: str) -> tuple[str, list[str]]:
    """Ask the chat model to answer ``topic`` from the rendered outline.

    Returns ``(answer, relevant_entity_names)`` — ``answer`` is empty unless the
    model judged the outline sufficient; the names are always returned so the
    caller can pull the underlying source chunks.
    """
    verdict = {}
    try:
        from rag.prompts.generator import form_message, message_fit_in

        user = f"Question:\n{topic}\n\n{noun.capitalize()}:\n{_render_structure(entities, relations)}\n\nOutput JSON:"
        _, msg = message_fit_in(form_message(_NAV_SYSTEM.format(noun=f"the {noun}"), user), tools.chat_mdl.max_length)
        ans = await tools.chat_mdl.async_chat(msg[0]["content"], msg[1:], {"temperature": 0.2})
        if isinstance(ans, tuple):
            ans = ans[0]
        cleaned = re.sub(r"^.*</think>", "", ans, flags=re.DOTALL)
        cleaned = re.sub(r"```(?:json)?\s*|\s*```", "", cleaned).strip()
        verdict = json_repair.loads(cleaned) or {}
        if not isinstance(verdict, dict):
            verdict = {}
    except Exception:
        _LOG.exception(f"[{label}] Could not read the outline with the model.")

    is_sufficient = bool(verdict.get("is_sufficient"))
    raw_answer = verdict.get("answer")
    answer = str(raw_answer).strip() if is_sufficient and raw_answer is not None else ""
    relevant = [n for n in (verdict.get("relevant_entities") or []) if isinstance(n, str)]
    _LOG.info(
        "[%s] The %s %s the question; %d relevant entity(ies): %s",
        label,
        noun,
        "answers" if is_sufficient else "does not fully answer",
        len(relevant),
        ", ".join(relevant[:10]) or "none",
    )
    return answer, relevant

File: internal/rag/advanced_rag/harness/structure_qa.go
//
//  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

package harness

import (
	"context"
	"fmt"
	"strings"
	"unicode"

	"github.com/cloudwego/eino/schema"

	"ragflow/internal/agent/chat"
)

// Ask the chat model to answer a question from a compiled-structure outline.
//
// Mirrors Python harness/structure_qa.py. Both Go navigation paths that read
// compiled rows render entities + relations into the same compact outline, ask
// the model whether that outline alone answers the question, and use the
// returned relevant_entities to pull the underlying source chunks even when the
// outline is NOT sufficient (so evidence still flows back to the caller).
//
// Lives apart from navigation.go because it is pure prompt/render work with no
// store access, and because both callers would otherwise import it from a file
// that in turn imports them.

// structureQATemperature pins Python structure_qa.py:79's
// {"temperature": 0.2} for the outline-verdict call.
const structureQATemperature = 0.2

// RenderStructure renders a compiled structure (entities + relations) as a
// compact outline for the prompt. Mirrors Python _render_structure:
//
//	Entities:
//	  - name (type): description
//	(blank line)
//	Relations:
//	  - src -[type]-> tgt
//
// Both lists are capped at maxStructureEntities / maxStructureRelations; empty
// names and empty relation endpoints are dropped. Entities with no type fall
// back to "other"; relations with no type fall back to "related".
func RenderStructure(entities, relations []map[string]any) string {
	var lines []string
	if len(entities) > 0 {
		lines = append(lines, "Entities:")
		n := len(entities)
		if n > maxStructureEntities {
			n = maxStructureEntities
		}
		for _, e := range entities[:n] {
			name := strings.TrimSpace(strAny(e["name"]))
			if name == "" {
				continue
			}
			typ := strings.TrimSpace(orStr(strAny(e["type"]), "other"))
			desc := strings.Join(strings.Fields(strAny(e["description"])), " ")
			line := "- " + name + " (" + typ + ")"
			if desc != "" {
				line += ": " + desc
			}
			lines = append(lines, line)
		}
	}
	if len(relations) > 0 {
		lines = append(lines, "\nRelations:")
		n := len(relations)
		if n > maxStructureRelations {
			n = maxStructureRelations
		}
		for _, r := range relations[:n] {
			src := strings.TrimSpace(strAny(r["from"]))
			tgt := strings.TrimSpace(strAny(r["to"]))
			if src == "" || tgt == "" {
				continue
			}
			rt := strings.TrimSpace(orStr(strAny(r["type"]), "related"))
			lines = append(lines, "- "+src+" -["+rt+"]-> "+tgt)
		}
	}
	return strings.Join(lines, "\n")
}

// AskStructure asks the chat model to answer `topic` from the rendered outline.
//
// Mirrors Python _ask_structure. Returns (answer, relevant_entity_names):
// answer is empty unless the model judged the outline sufficient; the names are
// always returned so the caller can pull the underlying source chunks. `noun`
// is the display noun ("catalog" / "mindmap" / "knowledge graph"); `label` is
// the log tag. `model` is the request-scoped chat model (Python passes
// tools.chat_mdl); a nil model means no model is available and the call skips,
// mirroring Python's no-op when tools.chat_mdl is absent. Never raises: on any
// chat/parse failure both results are empty, matching Python's except-return of
// an empty verdict.
func AskStructure(ctx context.Context, model SessionModel, topic, noun, label string, entities, relations []map[string]any) (string, []string) {
	if model == nil {
		_LOG.Printf("[%s] structure QA skipped (no chat model)", label)
		return "", nil
	}
	system := strings.ReplaceAll(navSystemPrompt, "{noun}", "the "+noun)
	rendered := RenderStructure(entities, relations)
	user := fmt.Sprintf("Question:\n%s\n\n%s:\n%s\n\nOutput JSON:", topic, capitalizeWord(noun), rendered)

	// Mirror Python _ask_structure:
	//   message_fit_in(form_message(system, user), tools.chat_mdl.max_length)
	// tools.chat_mdl.max_length is exposed via ContextLengthModel; when absent,
	// the 8192 default (chat EffectiveContextLength) applies, matching Python's
	// LLM.max_length defaulting when the model config omits max_tokens.
	budget := 0
	if cl, ok := model.(ContextLengthModel); ok {
		budget = cl.ContextLength()
	}
	if budget <= 0 {
		budget = chat.EffectiveContextLength(0)
	}
	fitted, fitErr := chat.FitMessages(system, []schema.Message{
		*schema.UserMessage(user),
	}, budget)
	if fitErr != "" {
		_LOG.Printf("[%s] prompt fitting failed: %s", label, fitErr)
		return "", nil
	}
	// FitMessages may prepend/trim a system message; re-extract it so the model
	// call is exactly [system, user...] as Python sends msg[0] then msg[1:].
	if len(fitted) > 0 && fitted[0].Role == schema.System {
		system = fitted[0].Content
		fitted = fitted[1:]
	}
	msgs := make([]schema.Message, 0, 1+len(fitted))
	msgs = append(msgs, *schema.SystemMessage(system))
	msgs = append(msgs, fitted...)
	// Python hardcodes {"temperature": 0.2} for this node (structure_qa.py:79)
	// — a stable verdict on a mechanical render, so it must not sample hot. The
	// value is a pinned constant: every production carrier implements
	// TemperatureModel (compile-time assertion on InvokerSessionModel), so the
	// temperature is always sent; a carrier without per-call temperature
	// support falls back to its own default (Go-only provider limitation).
	var resp *ModelReply
	var err error
	if tm, ok := model.(TemperatureModel); ok {
		resp, err = tm.CompleteWithTemperature(ctx, msgs, nil, structureQATemperature)
	} else {
		_LOG.Printf("[%s] model %T cannot carry per-call temperature; using its default (Python would send %v)", label, model, structureQATemperature)
		resp, err = model.Complete(ctx, msgs, nil)
	}
	if err != nil {
		_LOG.Printf("[%s] could not read the outline with the model: %v", label, err)
		return "", nil
	}

	var verdict structureNavVerdict
	if err := UnmarshalModelJSON(resp.Content, &verdict); err != nil {
		_LOG.Printf("[%s] could not parse the outline verdict: %v", label, err)
		return "", nil
	}

	answer := ""
	if verdict.IsSufficient {
		answer = strings.TrimSpace(verdict.Answer)
	}
	relevant := make([]string, 0, len(verdict.RelevantEntities))
	for _, n := range verdict.RelevantEntities {
		if s := strings.TrimSpace(n); s != "" {
			relevant = append(relevant, s)
		}
	}
	outcome := "does not fully answer"
	if verdict.IsSufficient {
		outcome = "answers"
	}
	_LOG.Printf("[%s] the %s %s the question; %d relevant entity(ies)", label, noun, outcome, len(relevant))
	return answer, relevant
}

// strAny reads a map field as its plain string form.
func strAny(v any) string {
	switch x := v.(type) {
	case string:
		return x
	case fmt.Stringer:
		return x.String()
	case nil:
		return ""
	default:
		return fmt.Sprint(x)
	}
}

// capitalizeWord mirrors Python str.capitalize(): uppercase the first rune and
// lowercase every remaining rune. Python lowercases the rest, so e.g.
// "hELLo".capitalize() -> "Hello"; the previous Go version left the tail
// untouched ("HELLO"), which diverged.
func capitalizeWord(s string) string {
	if s == "" {
		return s
	}
	rs := []rune(s)
	rs[0] = unicode.ToUpper(rs[0])
	for i := 1; i < len(rs); i++ {
		rs[i] = unicode.ToLower(rs[i])
	}
	return string(rs)
}
